### DISCUSSION TITLE
Do I give up on coding or can I actually properly write software?

### DISCUSSION THREAD

**Discussion by ghost** [community/community#141829] [Programming Help]
### Body

```cpp
/*
This software can be freely modified -- Please feel free to add any features or bug fixes
AUTHOR: WorldsDumbestCoder
GitHub: https://github.com/WorldsDumbestCoder
*/
// Today is Monday October 14th, 2024
#include <map>
#include <iostream>
//#include <fstream>
#include <string>
#include <algorithm>

/**
 * This class provides the mapping and functions for/of the intermediate representation (IR) to assembly language.
 * It contains three maps, `x64`, `x32`, and `x16`, that map human-readable instruction names to their corresponding assembly language instructions.
 * The class provides three functions, `get_x16instruction()`, `get_x32instruction()`, and `get_x64instruction()`, that return the assembly language instruction for a given human-readable instruction name, if it is valid.
 * The class is designed to reduce function call overhead by inlining all of its code.
 * By x16, x32 and x64, I am referring to the 16 bit, 32 bit and 64 bit x86 architectures.
 * --WorldsDumbestCoder
 */
class IR{
private:



    // I am inlining all of my code so then I can maintain reduce the function call overhead.
    // And the reason why I am using std:: over using namespace std is because then I don't import the whole entire standard library's namespace which I don't think is
    // needed

    // This is the human readable mapping of x64 assembly -- WorldsDumbestCoder
    std::map<std::string, std::string>x64 = {
        {"ACCUMULATOR", "RAX"},
        {"BASE", "RBX"},
        {"COUNTER", "RCX"},
        {"STACK", "RSP"},
        {"DATA", "RDX"},
        {"COMPARE", "CMP"},
        {"JUMP", "JMP"},
        {"JUMPLESS", "JL"},
        {"JUMPEQUAL", "JE"},
        {"JUMPIFNOTEQUAL", "JNE"},
        {"JUMPIFGREATER", "JG"},
        {"JUMPIFNOTGREATER", "JNG"},
        {"PUSH", "PUSH"},
        {"POP", "POP"},
        {"MOVE", "MOV"},
    };
    



    // This is the human readable mapping of x32 assembly -- WorldsDumbestCoder
    std::map<std::string, std::string>x32 = {
        {"ACCUMULATOR", "EAX"},
        {"BASE", "EBX"},
        {"COUNTER", "ECX"},
        {"DATA", "EDX"},
        {"STACK", "ESP"},
        {"COMPARE", "CMP"},
        {"PUSH", "PUSH"},
        {"POP", "POP"},
        {"MOVE", "MOV"},
    };
    



    // This is the human readable mapping of x16 ASM -- WorldsDumbestCoder
    std::map<std::string, std::string>x16 = {
        {"ACCUMULATOR", "AX"},
        {"ACCUMULATORLOW", "AL"},
        {"ACCUMULATORHIGH", "AH"},
        {"HIGHSTACK", "SH"},
        {"LOWSTACK", "SL"},
        {"STACK", "SP"},
        {"COMPARE", "CMP"},
        {"JUMP", "JMP"},
        {"BASE", "BX"},
        {"LOWBASE", "BL"},
        {"HIGHBASE", "BH"},
        {"HIGHCOUNTER", "CH"},
        {"LOWCOUNTER", "CL"},
        {"HIGHDATA", "DH"},
        {"LOWDATA", "DL"},
        {"PUSH", "PUSH"},
        {"POP", "POP"},
        {"MOVE", "MOV"}
    };



public:
    // Any ideas for anything better than x16, x32 and x64? -- WorldsDumbestCoder
    
    // This function returns the equivalent x16 assembly instruction if the parameter `instruction` is a valid instruction -- WorldsDumbestCoder

    inline std::string get_x16instruction(std::string instruction){
        // This here checks if `instruction` is in x16
        if(this->x16.find(instruction) != this->x16.end()){
            return this->x16[instruction] + "\n";
        }else{
            return "Invalid x16 instruction \n";
        }
        return "";
    }

                                                                                            


    // This function returns the x32 IR equivalent of the parameter `instruction` if the instruction is a valid instruction -- WorldsDumbestCoder
    inline std::string get_x32instruction(std::string instruction){
        if(this->x32.find(instruction) != this->x32.end()){
            return this->x32[instruction] + "\n";   // This will return the x64 equivalent -- WorldsDumbestCoder
        }else{
            return "Invalid x32 instruction!\n";    
        }
        return "";
    }


    // This function returns the equivalent x64 instruction if the parameter `instruction` is valid. -- WorldsDumbestCoder
    inline std::string get_x64instruction(std::string instruction){
        if(this->x64.find(instruction) != this->x64.end()){
            return this->x64[instruction] + "\n";   // This will return the x64 assembly equivalent -- WorldSDumbestCoder
        }else{
           return "Invalid x64 instruction!\n";
        }
        return "";
    }

    inline std::string reverse_x16_map (std::string instruction){
        if(this->x16.find(instruction) != this->x64.end()){
            return this->x16[instruction] + "\n";
        }
    }
};




int main(){
    IR ir;
    /*
    std::fstream f;
    std::string fname;
    std::cout << "Enter the file name > ";
    std::cin >> fname;
    if(f){
    }else{
        std::cout << "The file " << fname << " does not exist!\n";
    }
    f.close();
    */
    std::cout << ir.get_x64instruction("POP"); 
    return 0;
}
```

This is written from scratch by me and I did not use any search engine, stackoverflow, GitHub Community, Quora etc.

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by Unmesh100** [comment]
From looking at your code, I can confidently say that you should absolutely continue coding! Let me explain why, and then offer some constructive suggestions:
Positive aspects of your code:

You have a clear understanding of important C++ concepts:

Classes and object-oriented programming
STL containers (std::map)
String handling
Basic file operations (though commented out)
Access modifiers (public/private)


You show good programming practices:-

Detailed comments explaining your code
Consistent formatting
Thoughtful namespace handling (explaining why you avoid using namespace std)
Documentation of your functions


You're thinking about performance (inline functions, function call overhead)
You're working on something complex and useful - an assembly instruction mapper


Key improvements in the revised version:

Error handling: Using exceptions instead of returning error strings
Const correctness: Maps are marked const since they don't change
Code reuse: Added a helper function to reduce duplication
Parameter passing: Using const references for strings
Complete reverse lookup: Added a working reverse lookup function
Memory efficiency: Removed unnecessary returns and using references
Exception safety: Added try-catch in main()

You're definitely on the right track! The fact that you wrote this from scratch without external references shows you have a good grasp of programming concepts. My suggestions are just refinements to make your code more robust and maintainable.



Improved IR Mapper Implementation:-
------------------------------------------

#include <map>
#include <iostream>
#include <string>
#include <algorithm>
#include <stdexcept>

class IR {
private:
    // Using const maps since they won't be modified after initialization
    const std::map<std::string, std::string> x64 = {
        {"ACCUMULATOR", "RAX"},
        {"BASE", "RBX"},
        {"COUNTER", "RCX"},
        {"STACK", "RSP"},
        {"DATA", "RDX"},
        {"COMPARE", "CMP"},
        {"JUMP", "JMP"},
        {"JUMPLESS", "JL"},
        {"JUMPEQUAL", "JE"},
        {"JUMPIFNOTEQUAL", "JNE"},
        {"JUMPIFGREATER", "JG"},
        {"JUMPIFNOTGREATER", "JNG"},
        {"PUSH", "PUSH"},
        {"POP", "POP"},
        {"MOVE", "MOV"},
    };

    const std::map<std::string, std::string> x32 = {
        {"ACCUMULATOR", "EAX"},
        {"BASE", "EBX"},
        {"COUNTER", "ECX"},
        {"DATA", "EDX"},
        {"STACK", "ESP"},
        {"COMPARE", "CMP"},
        {"PUSH", "PUSH"},
        {"POP", "POP"},
        {"MOVE", "MOV"},
    };

    const std::map<std::string, std::string> x16 = {
        {"ACCUMULATOR", "AX"},
        {"ACCUMULATORLOW", "AL"},
        {"ACCUMULATORHIGH", "AH"},
        {"HIGHSTACK", "SH"},
        {"LOWSTACK", "SL"},
        {"STACK", "SP"},
        {"COMPARE", "CMP"},
        {"JUMP", "JMP"},
        {"BASE", "BX"},
        {"LOWBASE", "BL"},
        {"HIGHBASE", "BH"},
        {"HIGHCOUNTER", "CH"},
        {"LOWCOUNTER", "CL"},
        {"HIGHDATA", "DH"},
        {"LOWDATA", "DL"},
        {"PUSH", "PUSH"},
        {"POP", "POP"},
        {"MOVE", "MOV"}
    };

    // Helper function to reduce code duplication
    std::string get_instruction(const std::map<std::string, std::string>& map, 
                              const std::string& instruction,
                              const std::string& arch) {
        auto it = map.find(instruction);
        if (it != map.end()) {
            return it->second + "\n";
        }
        throw std::invalid_argument("Invalid " + arch + " instruction: " + instruction);
    }

public:
    std::string get_x16instruction(const std::string& instruction) {
        return get_instruction(x16, instruction, "x16");
    }

    std::string get_x32instruction(const std::string& instruction) {
        return get_instruction(x32, instruction, "x32");
    }

    std::string get_x64instruction(const std::string& instruction) {
        return get_instruction(x64, instruction, "x64");
    }

    // Reverse lookup function that works for all architectures
    std::string reverse_lookup(const std::string& assembly_instruction) {
        // Helper lambda to search a map for a value and return the corresponding key
        auto find_in_map = [](const std::map<std::string, std::string>& m, 
                             const std::string& value) -> std::string {
            for (const auto& pair : m) {
                if (pair.second == value) {
                    return pair.first;
                }
            }
            return "";
        };

        // Try each map in turn
        std::string result = find_in_map(x16, assembly_instruction);
        if (!result.empty()) return result + "\n";

        result = find_in_map(x32, assembly_instruction);
        if (!result.empty()) return result + "\n";

        result = find_in_map(x64, assembly_instruction);
        if (!result.empty()) return result + "\n";

        throw std::invalid_argument("Assembly instruction not found: " + assembly_instruction);
    }
};

int main() {
    IR ir;
    try {
        std::cout << "x64 POP instruction: " << ir.get_x64instruction("POP");
        std::cout << "Reverse lookup of RAX: " << ir.reverse_lookup("RAX");
    } catch (const std::invalid_argument& e) {
        std::cerr << "Error: " << e.what() << std::endl;
    }
    return 0;
}

**Reply by ghost** [reply]
@Unmesh100,
What is const correctness going to do to help me and what if I use pointers to access the map, so then I can directly access the map over using variables increasing memory efficiency?

**Reply by ghost** [reply]
And I lied I had several issues doing the C++ code and I asked for help for the file I/O, but the rest is done by me by doing a lot of trial and error.
